src/blocks/libvirt.rs

```diff
@@ -0,0 +1,157 @@
+use virt::connect::Connect;
+
+use std::time::Duration;
+
+use crossbeam_channel::Sender;
+use serde_derive::Deserialize;
+
+use crate::blocks::{Block, ConfigBlock, Update};
+use crate::config::SharedConfig;
+use crate::de::deserialize_duration;
+use crate::errors::*;
+use crate::formatting::value::Value;
+use crate::formatting::FormatTemplate;
+use crate::scheduler::Task;
+use crate::widgets::text::TextWidget;
+use crate::widgets::I3BarWidget;
+use crate::widgets::State;
+
+pub struct Libvirt {
+    id: usize,
+    text: TextWidget,
+    format: FormatTemplate,
+    update_interval: Duration,
+    qemu_conn: Connect,
+}
+
+#[derive(Deserialize, Debug, Clone)]
+#[serde(deny_unknown_fields, default)]
+pub struct LibvirtConfig {
+    /// Update interval in seconds
+    #[serde(deserialize_with = "deserialize_duration")]
+    pub interval: Duration,
+
+    /// Format override
+    pub format: FormatTemplate,
+
+    /// URL to QEMU
+    pub qemu_url: String,
+}
+
+impl Default for LibvirtConfig {
+    fn default() -> Self {
+        Self {
+            interval: Duration::from_secs(5),
+            format: FormatTemplate::default(),
+            qemu_url: "qemu:///system".to_string(),
+        }
+    }
+}
+
+impl ConfigBlock for Libvirt {
+    type Config = LibvirtConfig;
+
+    fn new(
+        id: usize,
+        block_config: Self::Config,
+        shared_config: SharedConfig,
+        _: Sender<Task>,
+    ) -> Result<Self> {
+        let text = TextWidget::new(id, 0, shared_config)
+            .with_text("vms")
+            .with_icon("virtual-machine")?;
+
+        let qemu_conn = match Connect::open_read_only(&block_config.qemu_url)
+            .block_error("virt", "error when connecting to libvirt")
+        {
+            Ok(c) => c,
+            Err(e) => {
+                return Err(e);
+            }
+        };
+
+        Ok(Libvirt {
+            id,
+            text,
+            format: block_config.format.with_default("{running}")?,
+            update_interval: block_config.interval,
+            qemu_conn,
+        })
+    }
+}
+
+impl Block for Libvirt {
+    fn update(&mut self) -> Result<Option<Update>> {
+        // Check if the connection is still active and re-instatiate it if it's not
+        if !self
+            .qemu_conn
+            .is_alive()
+            .block_error("virt", "error when getting connection status to libvirt")?
+        {
+            self.qemu_conn = Connect::open_read_only(
+                &self
+                    .qemu_conn
+                    .get_uri()
+                    .block_error("virt", "error in retrieving URI information from libvirt")?,
+            )
+            .block_error("virt", "error in reconnecting to libvirt socket")?;
+        };
```

Comment:
I tried moving this to a seperate `Impl` block under libvirt, however I'm not smart enough to make it work with the ownership semantics of `self`.

Basically, this would be a good thing to run before each of our collections, since the collections take a few milliseconds to run and libvirt could become unavailable during that time.

Reply:
Can you share your attempt here?

Reply:
```rust
impl Libvirt {
    fn check_and_reconnect(&self) -> Result<Self> {
        match self
            .qemu_conn
            .is_alive()
            .block_error("virt", "error when getting connection status to libvirt")
        {
            Ok(true) => {}
            Ok(false) => {
                new = self.clone();
                new.qemu_conn = Connect::open_read_only(
                    &self
                        .qemu_conn
                        .get_uri()
                        .block_error("virt", "error in retrieving URI information from libvirt")
                        .unwrap(), /* Only happens if memory corruption; then we should panic anyway */
                )
                .block_error("virt", "error in reconnecting to libvirt socket")?;
                return Ok(new);
            }
            Err(e) => {
                return Err(e);
            }
        };
    }
}
```

Reply:
You can try
1) Take `&mut self` as an argument instead.
2) Just mutate `self` instead of `clone()`ing.

Reply:
Which could then be called by `self = &mut Libvirt::check_and_reconnect(self)?;` or `self.check_and_reconect()` I hope.

Reply:
> `self.check_and_reconect()`

Yes :)

Reply:
oh, I made it work. Nevermind. :)